flink-runtime/src/main/java/org/apache/flink/runtime/operators/hash/ReduceHashTable.java

```diff
@@ -0,0 +1,1048 @@
+	 * otherwise inserts a new entry.
+	 *
+	 * (If there are multiple entries with the same key, then it will update one of them.)
+	 *
+	 * @param record The record to be processed.
+	 */
+	public void processRecordWithReduce(T record) throws Exception {
+		if (closed) {
+			return;
+		}
+
+		T match = prober.getMatchFor(record, reuse);
+		if (match == null) {
+			prober.insertAfterNoMatch(record);
+		} else {
+			// do the reduce step
+			T res = reducer.reduce(match, record);
+
+			// We have given reuse to the reducer UDF, so create new one if object reuse is disabled
+			if (!objectReuseEnabled) {
+				reuse = buildSideSerializer.createInstance();
+			}
+
+			prober.updateMatch(res);
+		}
+	}
+
+	/**
+	 * Searches the hash table for a record with the given key.
+	 * If it is found, then it is overridden with the specified record.
+	 * Otherwise, the specified record is inserted.
+	 * @param record The record to insert or to replace with.
+	 * @throws IOException (EOFException specifically, if memory ran out)
+     */
+	@Override
+	public void insertOrReplaceRecord(T record) throws IOException {
+		if (closed) {
+			return;
+		}
+
+		T match = prober.getMatchFor(record, reuse);
+		if (match == null) {
+			prober.insertAfterNoMatch(record);
+		} else {
+			prober.updateMatch(record);
+		}
+	}
+
+	/**
+	 * Inserts the given record into the hash table.
+	 * Note: this method doesn't care about whether a record with the same key is already present.
+	 * @param record The record to insert.
+	 * @throws IOException (EOFException specifically, if memory ran out)
+     */
+	@Override
+	public void insert(T record) throws IOException {
+		if (closed) {
+			return;
+		}
+
+		final int hashCode = MathUtils.jenkinsHash(buildSideComparator.hash(record));
+		final int bucket = hashCode & numBucketsMask;
+		final int bucketSegmentIndex = bucket >>> numBucketsPerSegmentBits; // which segment contains the bucket
+		final MemorySegment bucketSegment = bucketSegments[bucketSegmentIndex];
+		final int bucketOffset = (bucket & numBucketsPerSegmentMask) << bucketSizeBits; // offset of the bucket in the segment
+		final long firstPointer = bucketSegment.getLong(bucketOffset);
+
+		try {
+			final long newFirstPointer = recordArea.appendPointerAndRecord(firstPointer, record);
+			bucketSegment.putLong(bucketOffset, newFirstPointer);
+		} catch (EOFException ex) {
+			compactOrThrow();
+			insert(record);
+			return;
+		}
+
+		numElements++;
+		resizeTableIfNecessary();
+	}
+
+	private void resizeTableIfNecessary() throws IOException {
+		if (enableResize && numElements > numBuckets) {
+			final long newNumBucketSegments = 2L * bucketSegments.length;
+			// Checks:
+			// - we can't handle more than Integer.MAX_VALUE buckets
+			// - don't take more memory than the free memory we have left
+			// - the buckets shouldn't occupy more than half of all our memory
+			if (newNumBucketSegments * numBucketsPerSegment < Integer.MAX_VALUE &&
```

Comment:
That wouldn't work because the number of bucket segments has to be a power of 2, so that taking modulo is fast by bitwise and-ing with a mask, instead of dividing [1].

[1] https://gmplib.org/~tege/x86-timing.pdf